### ISSUE TITLE
pytorch code for Improved Training of Wasserstein GANs

### ISSUE THREAD

**Issue by aosokin** [martinarjovsky/WassersteinGAN#36]
Hi, do you plan to provide a pytorch implementation of [the recent paper](https://arxiv.org/abs/1704.00028) on "Improved Training of Wasserstein GANs"?
Is there an easy way to compute the gradient w.r.t. of the gradient norm?

**Comment by atgambardella** [comment]
I think on your input variable, you can just set requires_grad=True, and then take torch.norm(your_input.grad, 2)

**Comment by ccurro** [comment]
https://github.com/igul222/improved_wgan_training

**Comment by aosokin** [comment]
@atgambardella One needs the gradient of the gradient norm w.r.t. parameters, not the gradient w.r.t. the input

@ccurro This is a tensorflow code. I'm not aware of any pytorch implementation

**Comment by atgambardella** [comment]
@aosokin Section 4 of the paper, "Sampling across straight lines says":

> The gradient term ||∇xˆD(xˆ)||2 is with respect to the points xˆ, not the parameters of D

**Comment by aosokin** [comment]
@atgambardella agreed, my bad.
However this does not solve the problem. One needs to differentiate this norm, not simply compute it.

**Comment by soumith** [comment]
Hi Anton,

This cannot be implemented in pytorch until we merge the PR https://github.com/pytorch/pytorch/pull/1016 which enables gradients of gradients

**Comment by aosokin** [comment]
@soumith Thanks a lot for your reply! Looking forward to see that happen!

**Comment by NickShahML** [comment]
Is there a way to know when this merger has occurred?

**Comment by apaszke** [comment]
@NickShahML you can subscribe to the PR

**Comment by fonfonx** [comment]
Is it possible to implement it in Torch? How can we compute the gradient of the norm of the gradient of D?

**Comment by AjayTalati** [comment]
Might be possible to implement this in PyTorch now, the merge was done yesterday :+1:

**Comment by jfsantos** [comment]
This repository has an implementation of it using the new gradients of gradients feature: https://github.com/caogang/wgan-gp.

**Comment by rafaelvalle** [comment]
@jfsantos gradients of gradients? what do you mean by that?

**Comment by jfsantos** [comment]
@rafaelvalle For normalizing w.r.t. the norm of the gradients of the discriminator, you need to be able to backprop through that gradient, so you need to compute the gradient of a gradient. This is a recent feature in PyTorch's autograd.

**Comment by rafaelvalle** [comment]
@jfsantos when you say w.r.t. the norm of the gradients of the discriminator, do you mean w.r.t. to the points x_hat? I'm looking at page 6 of the paper.

**Comment by rafaelvalle** [comment]
@jfsantos yes! I got confused. w.r.t. the norm of the gradients of the discriminator on the interpolated points between Pr and Pg!

**Comment by LukasMosser** [comment]
@rafaelvalle @jfsantos but convolutions are not supported yet I believe.

**Comment by pclucas14** [comment]
Hi, 
does anyone know when convolutions will be supported ?

**Comment by jfsantos** [comment]
@pclucas14 Please follow this PR for info on double backward for convolutions: pytorch/pytorch#1643
